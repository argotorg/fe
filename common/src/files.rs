use crate::utils::keccak;
use crate::Span;
use codespan_reporting as cs;
use cs::files::Error as CsError;
use std::{
    collections::HashMap,
    convert::TryInto,
    fs,
    io,
    ops::Range,
    path::Path,
};

pub struct SourceFile {
    id: SourceFileId,
    name: String,
    content: String,
    line_starts: Vec<usize>,
}

#[derive(PartialEq, Copy, Clone, Eq, Hash)]
pub struct SourceFileId(u128);

impl SourceFile {
    pub fn new(name: String, content: String) -> Self {
        let hash = keccak::full_as_bytes(content.as_bytes());
        let line_starts = cs::files::line_starts(&content).collect();
        Self {
            id: SourceFileId(u128::from_be_bytes(hash[..16].try_into().unwrap())),
            name,
            content,
            line_starts,
        }
    }

    pub fn line_index(&self, byte_index: usize) -> usize {
        self.line_starts
            .binary_search(&byte_index)
            .unwrap_or_else(|next_line| next_line - 1)
    }

    pub fn line_span(&self, line_index: usize) -> Option<Span> {
        Some(Span::new(
            *self.line_starts.get(line_index)?,
            *self.line_starts.get(line_index + 1)?,
        ))
    }
}

pub trait FileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String>;
}

pub struct OsFileLoader;

impl FileLoader for OsFileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String> {
        fs::read_to_string(path)
    }
}

pub struct FileStore {
    files: HashMap<SourceFileId, SourceFile>,
    loader: Box<dyn FileLoader>,
}

impl FileStore {
    pub fn new() -> Self {
        Self {
            files: HashMap::new(),
            loader: Box::new(OsFileLoader),
        }
    }

    pub fn with_loader(loader: Box<dyn FileLoader>) -> Self {
        Self {
            files: HashMap::new(),
            loader,
        }
    }

    pub fn add_file(&mut self, path: String, content: String) -> SourceFileId {
        let file = SourceFile::new(path, content);
        let id = file.id;
        self.files.insert(id, file);
        id
    }

    pub fn load_file(&mut self, path: String) -> io::Result<(SourceFileId, String)> {
        let content = self.loader.load_file(&Path::new(&path))?;
        let id = self.add_file(path, content.clone());
        Ok((id, content))
    }

    pub fn get_file(&self, id: SourceFileId) -> Option<&SourceFile> {
        self.files.get(&id)
    }
}

impl<'a> cs::files::Files<'a> for FileStore {
    type FileId = SourceFileId;
    type Name = &'a str;
    type Source = &'a str;

    fn name(&'a self, id: SourceFileId) -> Result<Self::Name, CsError> {
        self.get_file(id)
            .map(|file| file.name.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn source(&'a self, id: SourceFileId) -> Result<Self::Source, CsError> {
        self.get_file(id)
            .map(|file| file.content.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn line_index(&'a self, id: SourceFileId, byte_index: usize) -> Result<usize, CsError> {
        Ok(self
            .get_file(id)
            .ok_or(CsError::FileMissing)?
            .line_index(byte_index))
    }

    fn line_range(&'a self, id: SourceFileId, line_index: usize) -> Result<Range<usize>, CsError> {
        let file = self.get_file(id).ok_or(CsError::FileMissing)?;
        Ok(file
            .line_span(line_index)
            .ok_or(CsError::LineTooLarge {
                given: line_index,
                max: file.line_starts.len() - 1,
            })?
            .into())
    }
}
