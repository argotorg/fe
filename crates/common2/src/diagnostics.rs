use parser::ast::SyntaxNodePtr;

use crate::InputFile;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct CompleteDiagnostic {
    pub severity: Severity,
    pub message: String,
    pub span: Span,
    pub sub_diagnostics: Vec<SubDiagnostic>,
    pub error_code: GlobalErrorCode,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct GlobalErrorCode {
    pub pass: AnalysisPass,
    pub local_code: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct SubDiagnostic {
    pub severity: Severity,
    pub message: String,
    pub span: Span,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Span {
    pub file: InputFile,
    pub node: SyntaxNodePtr,
    pub origin: SpanOrigin,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum Severity {
    Error,
    Warning,
    Note,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum SpanOrigin {
    Raw,
    Expanded,
    Desugared,
}

#[repr(u16)]
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum AnalysisPass {
    Parse = 1,
    NameResolution,
    TyCheck,

    ExternalAnalysis(ExternalAnalysisKey) = u16::MAX,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct ExternalAnalysisKey {
    name: String,
}
