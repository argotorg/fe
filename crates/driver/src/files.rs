use camino::Utf8PathBuf;

pub const FE_TOML: &str = "fe.toml";

pub fn find_project_root() -> Option<Utf8PathBuf> {
    let mut path = Utf8PathBuf::from_path_buf(
        std::env::current_dir().expect("Unable to get current directory"),
    )
    .expect("Expected utf8 path");

    loop {
        let fe_toml = path.join(FE_TOML);
        if fe_toml.is_file() {
            return Some(path);
        }

        if !path.pop() {
            break;
        }
    }

    None
}
