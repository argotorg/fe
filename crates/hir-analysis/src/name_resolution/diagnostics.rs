use hir::{
    hir_def::{IdentId, TopLevelMod},
    span::DynLazySpan,
};
use salsa::Update;
use thin_vec::ThinVec;

use super::NameRes;
use crate::HirAnalysisDb;

#[derive(Debug, Clone, PartialEq, Eq, Hash, Update)]
pub enum NameResDiag<'db> {
    /// The definition conflicts with other definitions.
    Conflict(IdentId<'db>, ThinVec<DynLazySpan<'db>>),

    /// The name is not found.
    NotFound(DynLazySpan<'db>, IdentId<'db>),

    /// The resolved name is not visible.
    Invisible(DynLazySpan<'db>, IdentId<'db>, Option<DynLazySpan<'db>>),

    /// The resolved name is ambiguous.
    Ambiguous(DynLazySpan<'db>, IdentId<'db>, Vec<DynLazySpan<'db>>),

    /// The name is found, but it can't be used as a middle segment of a path.
    InvalidPathSegment(DynLazySpan<'db>, IdentId<'db>, Option<DynLazySpan<'db>>),

    TooManyGenericArgs {
        span: DynLazySpan<'db>,
        expected: usize,
        given: usize,
    },

    /// The name is found but belongs to a different name domain other than the
    /// Type.
    ExpectedType(DynLazySpan<'db>, IdentId<'db>, &'static str),

    /// The name is found but belongs to a different name domain other than the
    /// trait.
    ExpectedTrait(DynLazySpan<'db>, IdentId<'db>, &'static str),

    /// The name is found but belongs to a different name domain other than the
    /// value.
    ExpectedValue(DynLazySpan<'db>, IdentId<'db>, &'static str),
}

impl<'db> NameResDiag<'db> {
    /// Returns the top-level module where the diagnostic is located.
    pub fn top_mod(&self, db: &'db dyn HirAnalysisDb) -> TopLevelMod<'db> {
        match self {
            Self::Conflict(_, conflicts) => conflicts
                .iter()
                .filter_map(|span| span.top_mod(db))
                .min()
                .unwrap(),
            Self::NotFound(span, _) => span.top_mod(db).unwrap(),
            Self::Invisible(span, _, _) => span.top_mod(db).unwrap(),
            Self::Ambiguous(span, _, _) => span.top_mod(db).unwrap(),
            Self::InvalidPathSegment(span, _, _) => span.top_mod(db).unwrap(),
            Self::ExpectedType(span, _, _) => span.top_mod(db).unwrap(),
            Self::ExpectedTrait(span, _, _) => span.top_mod(db).unwrap(),
            Self::ExpectedValue(span, _, _) => span.top_mod(db).unwrap(),
            Self::TooManyGenericArgs { span, .. } => span.top_mod(db).unwrap(),
        }
    }

    pub(super) fn ambiguous(
        db: &'db dyn HirAnalysisDb,
        span: DynLazySpan<'db>,
        ident: IdentId<'db>,
        cands: ThinVec<NameRes<'db>>,
    ) -> Self {
        let cands = cands
            .into_iter()
            .filter_map(|name| name.kind.name_span(db))
            .collect();
        Self::Ambiguous(span, ident, cands)
    }

    pub fn local_code(&self) -> u16 {
        match self {
            Self::Conflict(..) => 1,
            Self::NotFound(..) => 2,
            Self::Invisible(..) => 3,
            Self::Ambiguous(..) => 4,
            Self::InvalidPathSegment(..) => 5,
            Self::ExpectedType(..) => 6,
            Self::ExpectedTrait(..) => 7,
            Self::ExpectedValue(..) => 8,
            Self::TooManyGenericArgs { .. } => 9,
        }
    }
}
