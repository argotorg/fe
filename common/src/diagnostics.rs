use crate::files::{FileStore, SourceFileId};
use crate::Span;
pub use codespan_reporting::diagnostic::{
    Diagnostic as CsDiagnostic, Label as CsLabel, LabelStyle, Severity,
};
use codespan_reporting::term;
use term::termcolor::{BufferWriter, ColorChoice};

pub type Diagnostic = CsDiagnostic<SourceFileId>;

pub struct Label {
    pub style: LabelStyle,
    pub span: Span,
    pub message: String,
}
impl Label {
    /// Create a primary label with the given message. This will underline the
    /// given span with carets (`^^^^`).
    pub fn primary<S: Into<String>>(span: Span, message: S) -> Self {
        Label {
            style: LabelStyle::Primary,
            span,
            message: message.into(),
        }
    }

    /// Create a secondary label with the given message. This will underline the
    /// given span with hyphens (`----`).
    pub fn secondary<S: Into<String>>(span: Span, message: S) -> Self {
        Label {
            style: LabelStyle::Secondary,
            span,
            message: message.into(),
        }
    }

    /// Convert into a [`codespan_reporting::Diagnostic::Label`]
    pub fn into_cs_label(self, file_id: SourceFileId) -> CsLabel<SourceFileId> {
        CsLabel {
            style: self.style,
            file_id,
            range: self.span.into(),
            message: self.message,
        }
    }
}

/// Print the given diagnostics to stderr.
pub fn print_diagnostics(diagnostics: &[Diagnostic], files: &FileStore) {
    let writer = BufferWriter::stderr(ColorChoice::Auto);
    let mut buffer = writer.buffer();
    let config = term::Config::default();

    for diag in diagnostics {
        term::emit(&mut buffer, &config, files, &diag).unwrap();
    }
    // If we use `writer` here, the output won't be captured by rust's test system.
    eprintln!("{}", std::str::from_utf8(buffer.as_slice()).unwrap());
}

/// Format the given diagnostics as a string.
pub fn diagnostics_string(diagnostics: &[Diagnostic], files: &FileStore) -> String {
    let writer = BufferWriter::stderr(ColorChoice::Never);
    let mut buffer = writer.buffer();
    let config = term::Config::default();

    for diag in diagnostics {
        term::emit(&mut buffer, &config, files, &diag).expect("failed to emit diagnostic");
    }
    std::str::from_utf8(buffer.as_slice()).unwrap().to_string()
}
