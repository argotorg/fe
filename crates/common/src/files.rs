use crate::utils::keccak;
use crate::Span;
use codespan_reporting as cs;
use cs::files::Error as CsError;
use include_dir::Dir;
use indexmap::indexmap;
use indexmap::IndexMap;
use smol_str::SmolStr;
use std::collections::BTreeMap;
use std::ops::Range;
use std::path::Path;
use std::{fs, io};

#[derive(PartialEq, Clone, Eq, Hash, Debug)]
pub struct SourceFile {
    pub id: SourceFileId,
    pub name: String,
    pub content: String,
    line_starts: Vec<usize>,
}

#[derive(PartialEq, Copy, Clone, Eq, Hash, Debug, PartialOrd, Ord, Default)]
pub struct SourceFileId(pub u128);

impl SourceFile {
    pub fn new(name: &str, content: &str) -> Self {
        // Canonicalize new line character for Windows.
        let content_canonicalized = content.replace("\r\n", "\n");
        let hash = keccak::full_as_bytes(content_canonicalized.as_bytes());

        let line_starts = cs::files::line_starts(content).collect();
        Self {
            id: SourceFileId(u128::from_be_bytes(hash[..16].try_into().unwrap())),
            name: name.to_string(),
            content: content.to_string(),
            line_starts,
        }
    }

    pub fn line_index(&self, byte_index: usize) -> usize {
        self.line_starts
            .binary_search(&byte_index)
            .unwrap_or_else(|next_line| next_line - 1)
    }

    pub fn line_span(&self, line_index: usize) -> Option<Span> {
        let end = if line_index == self.line_starts.len() - 1 {
            self.content.len()
        } else {
            *self.line_starts.get(line_index + 1)?
        };
        Some(Span::new(self.id, *self.line_starts.get(line_index)?, end))
    }
}

pub trait FileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String>;
}

pub struct OsFileLoader;

impl FileLoader for OsFileLoader {
    fn load_file(&self, path: &Path) -> io::Result<String> {
        fs::read_to_string(path)
    }
}

pub struct FileStore {
    pub files: BTreeMap<SourceFileId, SourceFile>,
    loader: Box<dyn FileLoader>,
}

impl FileStore {
    pub fn new() -> Self {
        Self {
            files: BTreeMap::new(),
            loader: Box::new(OsFileLoader),
        }
    }

    pub fn with_loader(loader: Box<dyn FileLoader>) -> Self {
        Self {
            files: BTreeMap::new(),
            loader,
        }
    }

    pub fn add_file(&mut self, path: &str, content: &str) -> SourceFileId {
        let file = SourceFile::new(path, content);
        let id = file.id;
        self.files.insert(id, file);
        id
    }

    /// Adds an included dir to the file store.
    pub fn add_included_dir(&mut self, dir: &Dir) -> Vec<SourceFileId> {
        let mut file_ids = vec![];

        for file in dir.files() {
            file_ids.push(
                self.add_file(
                    file.path()
                        .to_str()
                        .expect("cannot convert file path to string"),
                    file.contents_utf8()
                        .expect("could not get utf8 encoded file content"),
                ),
            );
        }

        for sub_dir in dir.dirs() {
            file_ids.extend(self.add_included_dir(sub_dir))
        }

        file_ids
    }

    /// Adds the included libraries to the file store and returns a mapping of
    /// library names to file ids.
    pub fn add_included_libraries(&mut self) -> IndexMap<SmolStr, Vec<SourceFileId>> {
        indexmap! {
            "std".into() => self.add_included_dir(&fe_library::STD)
        }
    }

    pub fn load_file(&mut self, path: &str) -> io::Result<(String, SourceFileId)> {
        let content = self.loader.load_file(Path::new(&path))?;
        let id = self.add_file(path, &content);
        Ok((content, id))
    }

    pub fn get_file(&self, id: SourceFileId) -> Option<&SourceFile> {
        self.files.get(&id)
    }

    pub fn all_files(&self) -> Vec<SourceFileId> {
        self.files.keys().copied().collect()
    }
}

impl<'a> cs::files::Files<'a> for FileStore {
    type FileId = SourceFileId;
    type Name = &'a str;
    type Source = &'a str;

    fn name(&'a self, id: SourceFileId) -> Result<Self::Name, CsError> {
        self.get_file(id)
            .map(|file| file.name.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn source(&'a self, id: SourceFileId) -> Result<Self::Source, CsError> {
        self.get_file(id)
            .map(|file| file.content.as_str())
            .ok_or(CsError::FileMissing)
    }

    fn line_index(&'a self, id: SourceFileId, byte_index: usize) -> Result<usize, CsError> {
        Ok(self
            .get_file(id)
            .ok_or(CsError::FileMissing)?
            .line_index(byte_index))
    }

    fn line_range(&'a self, id: SourceFileId, line_index: usize) -> Result<Range<usize>, CsError> {
        let file = self.get_file(id).ok_or(CsError::FileMissing)?;
        Ok(file
            .line_span(line_index)
            .ok_or(CsError::LineTooLarge {
                given: line_index,
                max: file.line_starts.len() - 1,
            })?
            .into())
    }
}

impl Default for FileStore {
    fn default() -> Self {
        Self::new()
    }
}
