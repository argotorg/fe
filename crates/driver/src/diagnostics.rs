use std::ops::Range;

use camino::Utf8Path;
use codespan_reporting as cs;
use common::{
    diagnostics::{LabelStyle, Severity},
    InputDb, InputFile,
};
use cs::{diagnostic as cs_diag, files as cs_files};
use hir_analysis::diagnostics::{DiagnosticVoucher, SpannedHirAnalysisDb};

pub trait ToCsDiag {
    fn to_cs(&self, db: &dyn SpannedInputDb) -> cs_diag::Diagnostic<InputFile>;
}

pub trait SpannedInputDb: SpannedHirAnalysisDb + InputDb {}
impl<T> SpannedInputDb for T where T: SpannedHirAnalysisDb + InputDb {}

impl<T> ToCsDiag for T
where
    T: for<'db> DiagnosticVoucher<'db>,
{
    fn to_cs(&self, db: &dyn SpannedInputDb) -> cs_diag::Diagnostic<InputFile> {
        let complete = self.to_complete(db);

        let severity = convert_severity(complete.severity);
        let code = Some(complete.error_code.to_string());
        let message = complete.message;

        let labels = complete
            .sub_diagnostics
            .into_iter()
            .filter_map(|sub_diag| {
                let span = sub_diag.span?;
                match sub_diag.style {
                    LabelStyle::Primary => {
                        cs_diag::Label::new(cs_diag::LabelStyle::Primary, span.file, span.range)
                    }
                    LabelStyle::Secondary => {
                        cs_diag::Label::new(cs_diag::LabelStyle::Secondary, span.file, span.range)
                    }
                }
                .with_message(sub_diag.message)
                .into()
            })
            .collect();

        cs_diag::Diagnostic {
            severity,
            code,
            message,
            labels,
            notes: complete.notes,
        }
    }
}

fn convert_severity(severity: Severity) -> cs_diag::Severity {
    match severity {
        Severity::Error => cs_diag::Severity::Error,
        Severity::Warning => cs_diag::Severity::Warning,
        Severity::Note => cs_diag::Severity::Note,
    }
}

#[salsa::tracked(return_ref)]
pub fn file_line_starts(db: &dyn SpannedHirAnalysisDb, file: InputFile) -> Vec<usize> {
    cs::files::line_starts(file.contents(db).text(db)).collect()
}

pub struct CsDbWrapper<'a>(pub &'a dyn SpannedHirAnalysisDb);

impl<'db> cs_files::Files<'db> for CsDbWrapper<'db> {
    type FileId = InputFile;
    type Name = &'db Utf8Path;
    type Source = &'db str;

    fn name(&'db self, file_id: Self::FileId) -> Result<Self::Name, cs_files::Error> {
        Ok(file_id.path(self.0))
    }

    fn source(&'db self, file_id: Self::FileId) -> Result<Self::Source, cs_files::Error> {
        Ok(file_id.contents(self.0).text(self.0))
    }

    fn line_index(
        &'db self,
        file_id: Self::FileId,
        byte_index: usize,
    ) -> Result<usize, cs_files::Error> {
        let starts = file_line_starts(self.0, file_id);
        Ok(starts
            .binary_search(&byte_index)
            .unwrap_or_else(|next_line| next_line - 1))
    }

    fn line_range(
        &'db self,
        file_id: Self::FileId,
        line_index: usize,
    ) -> Result<Range<usize>, cs_files::Error> {
        let line_starts = file_line_starts(self.0, file_id);

        let start = *line_starts
            .get(line_index)
            .ok_or(cs_files::Error::LineTooLarge {
                given: line_index,
                max: line_starts.len() - 1,
            })?;

        let end = if line_index == line_starts.len() - 1 {
            file_id.contents(self.0).text(self.0).len()
        } else {
            *line_starts
                .get(line_index + 1)
                .ok_or(cs_files::Error::LineTooLarge {
                    given: line_index,
                    max: line_starts.len() - 1,
                })?
        };

        Ok(Range { start, end })
    }
}
