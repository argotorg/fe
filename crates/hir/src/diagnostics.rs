//! This module defines the diagnostics that can be accumulated inside salsa-db
//! with span-agnostic forms. All diagnostics accumulated in salsa-db should
//! implement [`DiagnosticVoucher`] which defines the conversion into
//! [`CompleteDiagnostics`].

use common::diagnostics::{CompleteDiagnostic, GlobalErrorCode};

use crate::span::db::SpannedHirDb;

/// All diagnostics accumulated in salsa-db should implement
/// [`DiagnosticVoucher`] which defines the conversion into
/// [`CompleteDiagnostic`].
///
/// All types that implement `DiagnosticVoucher` must NOT have a span
/// information which invalidates cache in salsa-db. Instead of it, the all
/// information is given by [`SpannedHirDB`] to allow evaluating span lazily.
///
/// The reason why we use `DiagnosticVoucher` is that we want to evaluate span
/// lazily to avoid invalidating cache in salsa-db.
///
/// To obtain a span from HIR nodes in a lazy manner, it's recommended to use
/// `[LazySpan]`(crate::span::LazySpan) and types that implement `LazySpan`.
pub trait DiagnosticVoucher: Send {
    fn error_code(&self) -> GlobalErrorCode;
    /// Consumes voucher and makes a [`CompleteDiagnostic`].
    fn to_complete(self, db: &dyn SpannedHirDb) -> CompleteDiagnostic;
}
