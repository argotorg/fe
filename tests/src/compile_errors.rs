//! Tests for contracts that should cause compile errors

use fe_common::diagnostics::{diagnostics_string, print_diagnostics};
use fe_common::files::FileStore;
use insta::assert_snapshot;

fn error_string(path: &str, src: &str) -> String {
    let mut files = FileStore::new();
    let id = files.add_file(path, src);

    let fe_module = match fe_parser::parse_file(src, id) {
        Ok((module, _)) => module,
        Err(diags) => {
            print_diagnostics(&diags, &files);
            panic!("parsing failed");
        }
    };
    let err = fe_analyzer::analyze(&fe_module, id).unwrap_err();

    let mut errstr = diagnostics_string(&err.diagnostics, &files);
    if let Some(classic) = err.classic {
        errstr.push_str(&format!("\n\n{}", classic.format_user(&src)));
    }
    errstr
}

macro_rules! test_file {
    ($name:ident) => {
        #[test]
        fn $name() {
            let path = format!("fixtures/compile_errors/{}.fe", stringify!($name));
            dbg!(&path);
            let src = std::fs::read_to_string(&path).unwrap();
            assert_snapshot!(error_string(&path, &src));
        }
    };
}

macro_rules! test_stmt {
    ($name:ident, $stmt:expr) => {
        #[test]
        fn $name() {
            let src = format!(
                "contract C:\n pub def f():\n  {}",
                $stmt.replace('\n', "\n  ")
            );
            assert_snapshot!(error_string("[snippet]", &src));
        }
    };
}

test_stmt! { array_non_primitive, "x: (u8, u8)[10]" }
test_stmt! { array_mixed_types, "x: u256[3] = [1, address(0), \"hi\"]" }
test_stmt! { assert_reason_not_string, "assert true, 1" }
test_stmt! { assign_int, "5 = 6" }
test_stmt! { assign_call, "self.f() = 10" }
test_stmt! { assign_type_mismatch, "x: u256 = 10\nx = address(0)" }
test_stmt! { aug_assign_non_numeric, "a: u256 = 1\nb: bool = true\na += b" }
test_stmt! { bad_map, "x: map<u8, u8, u8>" }
test_stmt! { bad_map2, "x: map<address, 100>" }
test_stmt! { bad_map3, "x: map<>" }
test_stmt! { bad_map4, "x: map<y>" }
test_stmt! { bad_map5, "x: map<map<u8, u8>, address>" }
test_stmt! { bad_map6, "x: map<10, 20>" }
test_stmt! { binary_op_add_uints, "a: u256 = 1\nb: u8 = 2\na + b" }
test_stmt! { binary_op_lshift_bool, "a: bool = true\nb: i256\na << b" }
test_stmt! { binary_op_lshift_with_int, "a: u256 = 1\nb: i256 = 2\na << b" }
test_stmt! { binary_op_pow_int, "a: u256 = 1\nb: i256 = 2\na ** b" }
test_stmt! { binary_op_boolean_mismatch1, "10 and true" }
test_stmt! { binary_op_boolean_mismatch2, "false or 1" }
test_stmt! { binary_op_boolean_mismatch3, "1 or 2" }
test_stmt! { break_without_loop, "break" }
test_stmt! { break_without_loop_2, "if true:\n  break" }
test_stmt! { call_undefined_function_on_contract, "self.doesnt_exist()" }
test_stmt! { clone_arg_count, "x: u256[2] = [5, 6]\ny: u256[2] = x.clone(y)" }
test_stmt! { continue_without_loop, "continue" }
test_stmt! { continue_without_loop_2, "if true:\n  continue" }
test_stmt! { emit_undefined_event, "emit MyEvent()" }
test_stmt! { keccak_called_with_wrong_type, "x: u256[1]\nkeccak256(foo=x, 10)" }
test_stmt! { non_bool_and, "x: bool = true\ny: u256 = 1\nx = x and y" }
test_stmt! { non_bool_or, "x: bool = true\ny: u256 = 1\nx = x or y" }
test_stmt! { overflow_i128_neg, "i128(-170141183460469231731687303715884105729)" }
test_stmt! { overflow_i128_pos, "i128(170141183460469231731687303715884105728)" }
test_stmt! { overflow_i16_neg, "i16(-32769)" }
test_stmt! { overflow_i16_pos, "i16(32768)" }
test_stmt! { overflow_i256_neg, "i256(-57896044618658097711785492504343953926634992332820282019728792003956564819969)" }
test_stmt! { overflow_i256_pos, "i256(57896044618658097711785492504343953926634992332820282019728792003956564819968)" }
test_stmt! { overflow_i32_neg, "i32(-2147483649)" }
test_stmt! { overflow_i32_pos, "i32(2147483648)" }
test_stmt! { overflow_i64_neg, "i64(-9223372036854775809)" }
test_stmt! { overflow_i64_pos, "i64(9223372036854775808)" }
test_stmt! { overflow_i8_neg, "i8(-129)" }
test_stmt! { overflow_i8_pos, "i8(128)" }
test_stmt! { overflow_literal_too_big, "115792089237316195423570985008687907853269984665640564039457584007913129639936" }
test_stmt! { overflow_literal_too_small, "-115792089237316195423570985008687907853269984665640564039457584007913129639936" }
test_stmt! { overflow_u128_neg, "u128(-1)" }
test_stmt! { overflow_u128_pos, "u128(340282366920938463463374607431768211456)" }
test_stmt! { overflow_u16_neg, "u16(-1)" }
test_stmt! { overflow_u16_pos, "u16(65536)" }
test_stmt! { overflow_u256_neg, "u256(-1)" }
test_stmt! { overflow_u256_pos, "u256(115792089237316195423570985008687907853269984665640564039457584007913129639936)" }
test_stmt! { overflow_u32_neg, "u32(-1)" }
test_stmt! { overflow_u32_pos, "u32(4294967296)" }
test_stmt! { overflow_u64_neg, "u64(-1)" }
test_stmt! { overflow_u64_pos, "u64(18446744073709551616)" }
test_stmt! { overflow_u8_neg, "u8(-1)" }
test_stmt! { overflow_u8_pos, "u8(256)" }
test_stmt! { pow_with_signed_exponent, "base: i128\nexp: i128\nbase ** exp" }
// Exponent can be unsigned but needs to be same size or smaller
test_stmt! { pow_with_wrong_capacity, "base: i128\nexp: u256\nbase ** exp" }
test_stmt! { string_capacity_mismatch, "String<3>(\"too long\")" }
test_stmt! { ternary_type_mismatch, "10 if 100 else true" }
test_stmt! { type_constructor_from_variable, "x: u8\ny: u16 = u16(x)" }
test_stmt! { type_constructor_arg_count, "x: u8 = u8(1, 10)" }
test_stmt! { unary_minus_on_bool, "x: bool = true\n-x" }
test_stmt! { unary_not_on_int, "x: u256 = 10\nnot x" }
test_stmt! { undefined_type, "x: foobar = 10" }
test_stmt! { undefined_generic_type, "x: foobar<u256> = 10" }
test_stmt! { unexpected_return, "return 1" }

test_file! { bad_tuple_attr1 }
test_file! { bad_tuple_attr2 }
test_file! { call_event_with_wrong_types }
test_file! { call_undefined_function_on_external_contract }
test_file! { call_undefined_function_on_memory_struct }
test_file! { call_undefined_function_on_storage_struct }
test_file! { circular_dependency_create }
test_file! { circular_dependency_create2 }
test_file! { duplicate_contract_in_module }
test_file! { duplicate_event_in_contract }
test_file! { duplicate_field_in_contract }
test_file! { duplicate_field_in_struct }
test_file! { duplicate_method_in_contract }
test_file! { duplicate_struct_in_module }
test_file! { duplicate_typedef_in_module }
test_file! { duplicate_var_in_child_scope }
test_file! { duplicate_var_in_contract_method }
test_file! { emit_bad_args }
test_file! { external_call_type_error }
test_file! { external_call_wrong_number_of_params }
test_file! { indexed_event }
test_file! { mismatch_return_type }
test_file! { missing_return }
test_file! { missing_return_in_else }
test_file! { needs_mem_copy }
test_file! { not_in_scope }
test_file! { not_in_scope_2 }
test_file! { return_addition_with_mixed_types }
test_file! { return_call_to_fn_with_param_type_mismatch }
test_file! { return_call_to_fn_without_return }
test_file! { return_from_init }
test_file! { return_lt_mixed_types }
test_file! { strict_boolean_if_else }
test_file! { struct_call_bad_args }
test_file! { struct_call_without_kw_args }
