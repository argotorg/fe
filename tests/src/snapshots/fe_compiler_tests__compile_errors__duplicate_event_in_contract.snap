---
source: tests/src/compile_errors.rs
expression: "error_string(&path, &src)"

---


AlreadyDefined on line 5
event MyEvent:
        idx addr1: address
